stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
   }
 }
 
+// Elements in `outputs` can come from two sources:
+// a) Symbolic tensors produced by tensor ops, and added as trace graph nodes.
+// b) Concrete tensors produced by host code (e.g. Tensor(1.0)).
+func finalizeAndExecuteTraceFn(fnName: String,
+                               inputs: [Tensor<Float>],
+                               outputs: [CTensorHandle]) -> [CTensorHandle] {
+  guard let tracedFn = _RuntimeConfig.traceState.getFunction else {
+    fatalError("Not in tracing mode!.")
+  }
+  _RuntimeConfig.traceState = .notTracing
+
+  return tracedFn.finalizeAndExecute(traceeBasicName: fnName,
+                                     inputs: inputs,
+                                     outputs: outputs)
+}
+
+public extension TensorArrayProtocol {
+  // The returned handles are owned by the caller.
+  func _getCTensorHandles() -> [CTensorHandle] {
+    debugLog("Getting \(self._tensorHandleCount) C handles.")
+    let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(
+      capacity: Int(self._tensorHandleCount))
+    debugLog("Unpacking handles into buffer.")
+    self._unpackTensorHandles(into: buffer)
+    let status = TF_NewStatus()
+    debugLog("Copying buffer content to output handles.")
+    var output: [CTensorHandle] = []
+    for idx in 0..<self._tensorHandleCount {
+      let address = buffer.advanced(by: Int(idx))
+      let isConcrete = TFE_TensorHandleIsConcrete(address.pointee) != 0
+      debugLog("""
+                 Copying the \(idx)-th C handle \(address.pointee) with \
+                 concrete=\(isConcrete).
+                 """)
+      let newHandle = TFE_TensorHandleCopySharingTensor(address.pointee,
+                                                        status)
+      checkOk(status)
+      debugLog("""
+                 Copying the \(idx)-th C handle \(address.pointee) with \
+                 concrete=\(isConcrete) to C handle create \(newHandle!).
+                 """)
+      output.append(newHandle!)
+    }
+    TF_DeleteStatus(status)
+    return output
+  }
+}
+
+public extension TensorGroup {
+  // The tensors in `input` are NOT owned by this instance.
+  init(_copying input: [CTensorHandle]) {
+    assert(Self._tensorHandleCount == input.count)
+    let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(capacity: input.count)
+    let status = TF_NewStatus()
+    // copy input to buffer
+    for (idx, inputTensorHandle) in input.enumerated() {
```

Comment:
`i`, if you care, but i don't have a strong opinion!
```suggestion
    for (i, inputTensorHandle) in input.enumerated() {
```

Reply:
Changed to `i`.